### ISSUE TITLE
[Suggestion] Tapping heroes to activate Astral Awakening

### ISSUE THREAD

**Issue by trant42** [Tune389/Tap-Titans-2-Progress-Bot#97]
Hi, I wonder if anyone is working on this feature. Apparently, CS uses AA in the build.
Hopefully, someone could share it!
Thanks.

**Comment by walterhuangio** [comment]
I don't know if AA is important for farming. Generally Coordinated Offensive is pretty good for farming and it's built into the script already.

**Comment by trant42** [comment]
I know that it's not really important, but tapping the heroes gives the huge buff for the damage since Hero damage is a part of main damage in CS build.

**Comment by FarSly** [comment]
@trant42 I have actually been working on this for the past couple days however its still not functional. The primary issue is color detection on the Astral Orb. If anyone has any ideas, please let me know! I'll share everything I have done thus far below.

I have tried both RGB and Color code detection with little success since the Orb is not a constant color. Also, it looks like the color debug is not 1px but rather an area in the cursor which is the problem imo. Its like it cant make up its mind on the color, so it just goes to -1 -1 -1 or -65793 a lot of the time. 

I have also tried 2 different approaches regarding location detection. My idea is to have it scan the screen in the areas where the Astral Orb will spawn, then click it if it's found. I believe this is the only way since clicking randomly would take far too long. You need the code to work its magic to provide an accurate click location.

**Attempt 1** 
Using a dual area "scan" detection. This will loop through the left and right side regions of the screen to find the Astral Orb color and then click it. This was my first attempt which I scrapped. I didnt really optimize it as it was just to get the basic functionality (which does work).

[astral.txt](https://github.com/Tune389/Tap-Titans-2-Progress-Bot/files/1911628/astral.txt)

**Attempt 2** 
This one is more in depth and has a full cycle imlemented but still isnt clicking correctly due to the color detection. I decided to take some sample coordinates from known Astral Orb locations and dump them into a list style. It will cycle the locations until a click, then cycle this 5 times before completing. The script works really well, but that damn color detection is failing me. 

[astral3.txt](https://github.com/Tune389/Tap-Titans-2-Progress-Bot/files/1911629/astral3.txt)

**Comment by chrisreyn** [comment]
`getColor` is an inefficient command for Hiro. it actually pauses your script for a split second which is noticeable. Adding this to the main script will very much delay your whole run by so much. I know because I used to use a script that heavily relied on `getColor` that the toasted time was super off

Also, `getColor` is not used to *look* for a color. it's used to *check* for a color. There is a difference.

You're better off coding a touch everywhere script than waste your run time on `getColor`

**Comment by Zethu5** [comment]
@chrisreyn  If you'll use a looping technique that will access a function every lets say 80~300 loops and will check AA than the lag that is created with getColor will by dramaticlly reduced.

**Comment by FarSly** [comment]
@chrisreyn Yeah, its clear this command is inefficient however what are the alternatives? I guess there are none other than blind clicking at random.

I can check ~30 locations a second with this but obviously, it's useless if nothing gets found accurately. 

I suppose this idea should be scrapped. Thanks for the comment!

**Comment by chrisreyn** [comment]
@FarSly tbh, if you noted down the locations of each hero, you could just do 5 loops of it instead of having to "make sure" it **is ** that hero with the glowing orb is all im saying. our only problem for now is when it goes to the equip/crate stack area (but that's also everyone's problem)
